### PR TITLE
Delete content from multiple users simultaneously

### LINKED ISSUE #3592: Delete content from multiple users simultaneously
There is already the option to delete the contents of a user. Using the clipboard, this action could also be applied to several users at the same time.

### PR DESCRIPTION
**Pull request by joshuaruesweg**
Close #3592

### CODE REVIEW

File: wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Ui/User/Content/Remove/Clipboard.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+   * Executes the remove content worker.
+   */
+  private executeWorker(objectTypes: string[]): void {
+    new AcpUiWorker({
+      // dialog
+      dialogId: "removeContentWorker",
+      dialogTitle: Language.get("wcf.acp.content.removeContent"),
+
+      // ajax
+      className: "\\wcf\\system\\worker\\UserContentRemoveWorker",
```
Comment (TimWolla) [inline wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Ui/User/Content/Remove/Clipboard.ts:65]:
```suggestion
      className: "wcf\\system\\worker\\UserContentRemoveWorker",
```

File: wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Ui/User/Content/Remove/Clipboard.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  public userIds: number[];
+  private readonly dialogId = "userContentRemoveClipboardPrepareDialog";
+
+  /**
+   * Initializes the content remove handler.
+   */
+  constructor() {
+    EventHandler.add("com.woltlab.wcf.clipboard", "com.woltlab.wcf.user", (data: EventData) => {
+      if (data.data.actionName === "com.woltlab.wcf.user.deleteUserContent") {
+        this.userIds = data.data.parameters.objectIDs;
```
Comment (TimWolla) [inline wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Ui/User/Content/Remove/Clipboard.ts:48]: Avoid state in the class. Pass the userIDs as the second parameter of `Ajax.api()`.

File: wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Ui/User/Content/Remove/Clipboard.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+  get dialogContent(): HTMLElement {
+    return UiDialog.getDialog(this.dialogId)!.content;
+  }
+
+  _ajaxSuccess(data: AjaxResponse): void {
+    UiDialog.open(this, data.returnValues.template);
+
+    const submitButton = this.dialogContent.querySelector('input[type="submit"]') as HTMLElement;
+    submitButton.addEventListener("click", () => this.submit());
```
Comment (TimWolla) [inline wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Ui/User/Content/Remove/Clipboard.ts:100]: Doesn't the Dialog itself provide callbacks for this?

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php [hunk 4]
```diff
@@ -29,9 +31,9 @@ class UserContentRemoveWorker extends AbstractWorker {
 	
 	/**
 	 * user
-	 * @var	User
+	 * @var	User[]
 	 */
-	protected $user = null;
+	protected $user = [];
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:36]: Rename this to `users`.

**Review by TimWolla — CHANGES_REQUESTED** [review]
Some first remarks. Generally I find the logic super hard to follow here and will need to spend more time understanding it.

File: wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Ui/User/Content/Remove/Clipboard.ts (hunk 3 above)
Comment (joshuaruesweg) [reply wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Ui/User/Content/Remove/Clipboard.ts:100]: No, unfortunately he does not. I wouldn't change it either, even if I would love to.

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php [hunk 5]
```diff
@@ -175,11 +203,21 @@ public function execute() {
...
 		$objectList->readObjectIDs();
 		$processor->deleteContent($objectList->objectIDs);
 		
+		$this->data['provider'][$providerObjectType]['user'][$userID]['count'] -= $this->limit;
+		
+		if ($this->data['provider'][$providerObjectType]['user'][$userID]['count'] <= 0) {
+			unset($this->data['provider'][$providerObjectType]['user'][$userID]);
+		} 
+		
 		$this->data['provider'][$providerObjectType]['count'] -= $this->limit;
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:221]: I believe this might be buggy with the refactoring now. Consider the following:

1. A limit of 10.
2. Two users with 3 posts each – a total of 6.
3. The first user is processed: This will subtract 10 from 6, resulting in -4, resulting in the second user now being processed.

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php [hunk 6]
```diff
@@ -128,19 +146,29 @@ private function generateData() {
...
+								'objectTypeID' => $contentProvider->objectTypeID,
+								'nicevalue' => $contentProvider->nicevalue ?: 0,
+								'user' => [],
+							];
+						}
+						
+						$this->data['provider'][$contentProvider->objectType]['user'][$user->userID] = [
+							'count' => $count,
+						];
+						$this->data['provider'][$contentProvider->objectType]['count'] += $count;
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:168]: This is a horrible data structure and I believe this is part of the reason the issue happened. May I suggest the following:

1. Sort the providers by niceValue before iterating them.
     - They are already in the correct order then.
2. Invert the 2 loops.
     - This processes all contents for a single user before processing the next user, feeling more natural.
3. Then make this a flat queue:
     ```
     $this->queue[] = [ 'userID' => …, 'provider' => …, 'remaining' => … ];
     ```

When processing you then simply take the first item in the queue and process it until `remaining < 0`, then removing it and proceeding with the next. No need for complex nesting. The number of remaining items then is a simple `array_sum(array_column($this->queue, 'remaining'))` away.

**Review by TimWolla — CHANGES_REQUESTED** [review]

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php (hunk 5 above)
Comment (joshuaruesweg) [reply wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:221]: No, because of the following code during the initialisation: https://github.com/WoltLab/WCF/blob/8ea6fb91cb5abd62470724f54c320d4df87e8ffe/wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php#L170

File: wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Ui/User/Content/Remove/Clipboard.ts (hunk 2 above)
Comment (joshuaruesweg) [reply wcfsetup/install/files/ts/WoltLabSuite/Core/Acp/Ui/User/Content/Remove/Clipboard.ts:48]: No I can't, because I need it during the execution of the worker.

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php (hunk 6 above)
Comment (joshuaruesweg) [reply wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:168]: I think it complexifies the logic way more, then it is now. Especially because we currently only execute the object types for the users that really have content (and object types w/o any content will be skipped anyway). In a flat map, I need to test that with every object type again. Keep in mind that the logic has been in use for 2 (?) versions.

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php [hunk 7]
```diff
@@ -29,9 +31,9 @@ class UserContentRemoveWorker extends AbstractWorker {
 	
 	/**
 	 * user
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:33]:
```suggestion
```

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php [hunk 8]
```diff
@@ -49,18 +51,34 @@ class UserContentRemoveWorker extends AbstractWorker {
...
-			throw new \InvalidArgumentException('userID is unknown.');
+		if (isset($this->parameters['userIDs']) && is_array($this->parameters['userIDs']) && !empty($this->parameters['userIDs'])) {
+			$userList = new UserList();
+			$userList->setObjectIDs($this->parameters['userIDs']);
+			$userList->readObjects();
+			
+			if ($userList->count() !== count($this->parameters['userIDs'])) {
+				$diff = array_diff($this->parameters['userIDs'], array_map(function (User $user) {
+					return $user->userID;
+				}, $userList->getObjects()));
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:66]:
```suggestion
				$diff = array_diff($this->parameters['userIDs'], array_column($userList->getObjects(), 'userID'));
```

Could apply likewise to `$knownContentProvider` below.

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php [hunk 9]
```diff
@@ -210,4 +252,16 @@ public function finalize() {
...
 		return LinkHandler::getInstance()->getLink('UserList');
 	}
+	
+	/**
+	 * Generates a key for session data saving.
+	 */
+	protected function generateKey(): string {
+		$userIDs = array_map(function (User $user) {
+			return $user->userID;
+		}, $this->users);
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:262]:
```suggestion
		$userIDs = array_column($this->users, 'userID');
```

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php (hunk 6 above)
Comment (TimWolla) [reply wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:168]: > Especially because we currently only execute the object types for the users that really have content (and object types w/o any content will be skipped anyway).

You are free to not put the entries into the queue if `count == 0`.

> Keep in mind that the logic has been in use for 2 (?) versions.

The previous logic has been in use for several versions, but the logic to handle multiple users is new.

And quite frankly: I'm having a very hard time following the control flow here due to `$this->data` containing 5 (!) nested arrays where you need to subtract the number of processed items in various locations to make sure to correctly remove specific parts of the arrays to indicate that this part was fully processed.

A queue is as simple of a data structure as you can get. It _will_ drastically simplify the code: Process the first `(userID, objectTypeID)` pair until `count` is non-positive, repeat until no more pairs exist.

**Review by TimWolla — CHANGES_REQUESTED** [review]

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php [hunk 10]
```diff
@@ -126,32 +135,36 @@ private function generateData() {
...
-				$contentList = $processor->getContentListForUser($this->user);
-				$count = $contentList->countObjects();
-				
-				if ($count) {
-					$this->data['provider'][$contentProvider->objectType] = [
-						'count' => $count,
-						'objectTypeID' => $contentProvider->objectTypeID,
-						'nicevalue' => $contentProvider->nicevalue ?: 0
-					];
+				foreach ($this->users as $user) {
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:150]: I would invert the order of the 2 loops:

Outer loop: Users.
Inner Loop: Content Providers.

This ensures that users after processed one after another which is the more natural order IMO.

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php [hunk 11]
```diff
@@ -126,32 +135,36 @@ private function generateData() {
 			}
 		}
 		
+		$contentProviders = ObjectTypeCache::getInstance()->getObjectTypes('com.woltlab.wcf.content.userContentProvider');
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:138]: This could be removed if `$this->contentProvider` would store the `$objectType` instead of the `$objectTypeName`. Then you could directly sort the actual list and could get rid of the condition within the loop.

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php [hunk 12]
```diff
@@ -103,8 +114,6 @@ private function generateData() {
 			'count' => 0
 		];
 		
-		$contentProviders = ObjectTypeCache::getInstance()->getObjectTypes('com.woltlab.wcf.content.userContentProvider');
-		
 		// add the required object types for the select content provider
 		if (is_array($this->contentProvider)) {
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:118]:
```suggestion
		if (is_array($this->contentProviders)) {
```
The variable should be renamed to be less confusing.

**Review by TimWolla — COMMENTED** [review]
Thanks. This is much cleaner now. I've some remark to make this a bit more readable, but the functionality LGTM.

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php (hunk 10 above)
Comment (joshuaruesweg) [reply wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:150]: I think in the end it doesn't matter. I want to avoid having to check the content providers several times. That would make the code less readable, because the user doesn't notice when and how what is deleted anyway.

File: wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php (hunk 11 above)
Comment (joshuaruesweg) [reply wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:138]: Keep in mind, that the class variable `$contentProvider` is a parameter, which is read in `validate()`. In the end, it doesn't matter where the whole thing is sorted (and does not clean up the code).

Comment (TimWolla) [reply wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php:137]: I was thinking about this (untested):

```
diff --git i/wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php w/wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php
index 1649009ed5..795cae338e 100644
--- i/wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php
+++ w/wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php
@@ -1,5 +1,6 @@
 <?php
 namespace wcf\system\worker;
+use wcf\data\object\type\ObjectType;
 use wcf\data\object\type\ObjectTypeCache;
 use wcf\data\user\User;
 use wcf\data\user\UserList;
@@ -113,6 +114,9 @@ class UserContentRemoveWorker extends AbstractWorker {
 			'count' => 0
 		];
 		
+		/** @var ObjectType[] $contentProviders */
+		$contentProviders = [];
+		
 		// add the required object types for the select content provider
 		if (is_array($this->contentProviders)) {
 			foreach ($this->contentProviders as $contentProvider) {
@@ -128,13 +132,14 @@ class UserContentRemoveWorker extends AbstractWorker {
 							throw new \RuntimeException('Unknown required object type "' . $objectTypeName . '" for object type "' . $contentProvider . '" given.');
 						}
 						
-						$this->contentProviders[] = $objectTypeName;
+						$contentProviders[] = $objectType;
 					}
 				}
 			}
 		}
-		
-		$contentProviders = ObjectTypeCache::getInstance()->getObjectTypes('com.woltlab.wcf.content.userContentProvider');
+		else {
+			$contentProviders = ObjectTypeCache::getInstance()->getObjectTypes('com.woltlab.wcf.content.userContentProvider');
+		}
 		
 		// sort object types
 		uasort($contentProviders, function ($a, $b) {
@@ -145,22 +150,20 @@ class UserContentRemoveWorker extends AbstractWorker {
 		});
 		
 		foreach ($contentProviders as $contentProvider) {
-			if ($this->contentProviders === null || (is_array($this->contentProviders) && in_array($contentProvider->objectType, $this->contentProviders))) {
-				foreach ($this->users as $user) {
-					/** @var IUserContentProvider $processor */
-					$processor = $contentProvider->getProcessor();
-					$contentList = $processor->getContentListForUser($user);
-					$count = $contentList->countObjects();
+			foreach ($this->users as $user) {
+				/** @var IUserContentProvider $processor */
+				$processor = $contentProvider->getProcessor();
+				$contentList = $processor->getContentListForUser($user);
+				$count = $contentList->countObjects();
+				
+				if ($count) {
+					$this->data['provider'][] = [
+						'userID' => $user->userID,
+						'objectTypeID' => $contentProvider->objectTypeID,
+						'count' => $count
+					];
 					
-					if ($count) {
-						$this->data['provider'][] = [
-							'userID' => $user->userID,
-							'objectTypeID' => $contentProvider->objectTypeID,
-							'count' => $count
-						];
-						
-						$this->data['count'] += ceil($count / $this->limit) * $this->limit;
-					}
+					$this->data['count'] += ceil($count / $this->limit) * $this->limit;
 				}
 			}
 		}
```

With Whitespace ignored:

```
diff --git i/wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php w/wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php
index 1649009ed5..795cae338e 100644
--- i/wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php
+++ w/wcfsetup/install/files/lib/system/worker/UserContentRemoveWorker.class.php
@@ -1,5 +1,6 @@
 <?php
 namespace wcf\system\worker;
+use wcf\data\object\type\ObjectType;
 use wcf\data\object\type\ObjectTypeCache;
 use wcf\data\user\User;
 use wcf\data\user\UserList;
@@ -113,6 +114,9 @@ class UserContentRemoveWorker extends AbstractWorker {
 			'count' => 0
 		];
 		
+		/** @var ObjectType[] $contentProviders */
+		$contentProviders = [];
+		
 		// add the required object types for the select content provider
 		if (is_array($this->contentProviders)) {
 			foreach ($this->contentProviders as $contentProvider) {
@@ -128,13 +132,14 @@ class UserContentRemoveWorker extends AbstractWorker {
 							throw new \RuntimeException('Unknown required object type "' . $objectTypeName . '" for object type "' . $contentProvider . '" given.');
 						}
 						
-						$this->contentProviders[] = $objectTypeName;
+						$contentProviders[] = $objectType;
 					}
 				}
 			}
 		}
-		
+		else {
 			$contentProviders = ObjectTypeCache::getInstance()->getObjectTypes('com.woltlab.wcf.content.userContentProvider');
+		}
 		
 		// sort object types
 		uasort($contentProviders, function ($a, $b) {
@@ -145,7 +150,6 @@ class UserContentRemoveWorker extends AbstractWorker {
 		});
 		
 		foreach ($contentProviders as $contentProvider) {
-			if ($this->contentProviders === null || (is_array($this->contentProviders) && in_array($contentProvider->objectType, $this->contentProviders))) {
 			foreach ($this->users as $user) {
 				/** @var IUserContentProvider $processor */
 				$processor = $contentProvider->getProcessor();
@@ -164,7 +168,6 @@ class UserContentRemoveWorker extends AbstractWorker {
 			}
 		}
 	}
-	}
 	
 	/**
 	 * @inheritDoc
```

Which I'd say *is* cleaner, because the `if` could be removed.

**Review by TimWolla — APPROVED** [review]
Overall this LGTM now.

File: wcfsetup/install/files/acp/templates/userList.tpl [hunk 13]
```diff
@@ -51,6 +51,8 @@
 			});
 			
 			new AcpUiUserList();
+			
+			new AcpUserContentRemoveClipboard();
```
Comment (TimWolla) [inline wcfsetup/install/files/acp/templates/userList.tpl:55]: Bad merge: The file is not listed in `require()`.

**Review by TimWolla — CHANGES_REQUESTED** [review]
